File: src/defined_names.cpp
#include <nanobind/stl/optional.h>

#include "bindings.hpp"

namespace {
struct XLDefinedNameIterator {
    std::vector<XLDefinedName> names;
    size_t index = 0;

    XLDefinedNameIterator(std::vector<XLDefinedName>&& n) : names(std::move(n)) {}

    XLDefinedName* next() {
        if (index < names.size()) return &names[index++];
        return nullptr;
    }
};
}  // namespace

void init_defined_names(py::module_& m) {
    // Bind XLDefinedNameIterator
    py::class_<XLDefinedNameIterator>(m, "XLDefinedNameIterator")
        .def("__next__",
             [](XLDefinedNameIterator& self) -> XLDefinedName {
                 auto* res = self.next();
                 if (!res) throw py::stop_iteration();
                 return *res;
             })
        .def("__iter__", [](XLDefinedNameIterator& self) { return self; });

    // Bind XLDefinedName
    py::class_<XLDefinedName>(m, "XLDefinedName")
        .def(py::init<>())
        .def("name", &XLDefinedName::name)
        .def("set_name", &XLDefinedName::setName)
        .def("refers_to", &XLDefinedName::refersTo)
        .def("set_refers_to", &XLDefinedName::setRefersTo)
        .def("local_sheet_id",
             [](const XLDefinedName& self) -> py::object {
                 auto val = self.localSheetId();
                 return val ? py::cast(*val) : py::none();
             })
        .def("set_local_sheet_id", &XLDefinedName::setLocalSheetId)
        .def("hidden", &XLDefinedName::hidden)
        .def("set_hidden", &XLDefinedName::setHidden)
        .def("comment", &XLDefinedName::comment)
        .def("set_comment", &XLDefinedName::setComment)
        .def("valid", &XLDefinedName::valid);

    // Bind XLDefinedNames
    py::class_<XLDefinedNames>(m, "XLDefinedNames")
        .def(
            "append",
            [](XLDefinedNames& self, std::string_view name, std::string_view formula) {
                return self.append(name, formula, std::nullopt);
            },
            py::arg("name"), py::arg("formula"))
        .def(
            "append",
            [](XLDefinedNames& self, std::string_view name, std::string_view formula,
               uint32_t localSheetId) { return self.append(name, formula, localSheetId); },
            py::arg("name"), py::arg("formula"), py::arg("local_sheet_id"))
        .def(
            "remove",
            [](XLDefinedNames& self, std::string_view name) { self.remove(name, std::nullopt); },
            py::arg("name"))
        .def(
            "remove",
            [](XLDefinedNames& self, std::string_view name, uint32_t localSheetId) {
                self.remove(name, localSheetId);
            },
            py::arg("name"), py::arg("local_sheet_id"))
        .def(
            "get",
            [](const XLDefinedNames& self, std::string_view name) {
                return self.get(name, std::nullopt);
            },
            py::arg("name"))
        .def(
            "get",
            [](const XLDefinedNames& self, std::string_view name, uint32_t localSheetId) {
                return self.get(name, localSheetId);
            },
            py::arg("name"), py::arg("local_sheet_id"))
        .def("all", &XLDefinedNames::all)
        .def(
            "exists",
            [](const XLDefinedNames& self, std::string_view name) {
                return self.exists(name, std::nullopt);
            },
            py::arg("name"))
        .def(
            "exists",
            [](const XLDefinedNames& self, std::string_view name, uint32_t localSheetId) {
                return self.exists(name, localSheetId);
            },
            py::arg("name"), py::arg("local_sheet_id"))
        .def("count", &XLDefinedNames::count)
        .def("__len__", &XLDefinedNames::count)
        .def("__getitem__",
             [](const XLDefinedNames& self, const std::string& name) {
                 auto val = self.get(name);
                 if (!val.valid())
                     throw py::key_error(("Defined name '" + name + "' not found").c_str());
                 return val;
             })
        .def(
            "__iter__",
            [](const XLDefinedNames& self) { return XLDefinedNameIterator(self.all()); },
            py::keep_alive<0, 1>());
}

File: src/pyopenxlsx/defined_names.py
"""High-level façade for workbook defined names (named ranges)."""

from __future__ import annotations

from typing import Any, Iterator, List, Optional


class DefinedName:
    """Thin wrapper around a native ``XLDefinedName``.

    Forwards attribute access to the native object (``name()``,
    ``refers_to()``, …) so existing call sites keep working.
    """

    __slots__ = ("_dn",)

    def __init__(self, native: Any):
        self._dn = native

    @property
    def raw(self) -> Any:
        return self._dn

    def __getattr__(self, name: str) -> Any:
        return getattr(self._dn, name)

    def __repr__(self) -> str:
        try:
            return f"DefinedName({self._dn.name()!r} -> {self._dn.refers_to()!r})"
        except Exception:
            return "DefinedName(<invalid>)"


class DefinedNames:
    """Collection API for workbook named ranges.

    Wraps native ``XLDefinedNames`` while keeping ``append`` / ``get`` /
    ``exists`` / ``remove`` compatible with existing call sites. Prefer
    :meth:`define` for new code.
    """

    __slots__ = ("_names", "_workbook")

    def __init__(self, native: Any, workbook: Any = None):
        self._names = native
        self._workbook = workbook

    @property
    def raw(self) -> Any:
        return self._names

    def define(
        self,
        name: str,
        refers_to: str,
        *,
        local_sheet_id: Optional[int] = None,
        sheet: Any = None,
        hidden: bool = False,
        comment: str = "",
    ) -> DefinedName:
        """Define (or redefine) a named range.

        :param name: Defined name (no spaces recommended).
        :param refers_to: Formula/reference, e.g. ``\"Sheet1!$A$1:$B$10\"``.
        :param local_sheet_id: Optional 0-based sheet-local scope.
        :param sheet: Alternative to *local_sheet_id*; a high-level
            ``Worksheet`` or sheet title string.
        :param hidden: Hide the name from Excel's UI when True.
        :param comment: Optional comment on the defined name.
        """
        sheet_id = local_sheet_id
        if sheet is not None:
            sheet_id = self._resolve_sheet_id(sheet)

        # Replace existing name in the same scope for idempotent defines.
        if sheet_id is None:
            if self._names.exists(name):
                self._names.remove(name)
            self._names.append(name, refers_to)
            dn = self._names.get(name)
        else:
            if self._names.exists(name, local_sheet_id=sheet_id):
                self._names.remove(name, local_sheet_id=sheet_id)
            self._names.append(name, refers_to, sheet_id)
            dn = self._names.get(name, local_sheet_id=sheet_id)

        if hidden:
            dn.set_hidden(True)
        if comment:
            dn.set_comment(comment)
        return DefinedName(dn)

    def append(
        self,
        name: str,
        formula: str,
        local_sheet_id: Optional[int] = None,
    ) -> Any:
        """Native-compatible append (does not replace existing names)."""
        if local_sheet_id is None:
            return self._names.append(name, formula)
        return self._names.append(name, formula, local_sheet_id)

    def get(
        self, name: str, local_sheet_id: Optional[int] = None
    ) -> DefinedName:
        if local_sheet_id is None:
            return DefinedName(self._names.get(name))
        return DefinedName(self._names.get(name, local_sheet_id=local_sheet_id))

    def remove(self, name: str, local_sheet_id: Optional[int] = None) -> None:
        if local_sheet_id is None:
            self._names.remove(name)
        else:
            self._names.remove(name, local_sheet_id=local_sheet_id)

    def exists(self, name: str, local_sheet_id: Optional[int] = None) -> bool:
        if local_sheet_id is None:
            return self._names.exists(name)
        return self._names.exists(name, local_sheet_id=local_sheet_id)

    def count(self) -> int:
        return self._names.count()

    def all(self) -> List[DefinedName]:
        return [DefinedName(dn) for dn in self._names.all()]

    def __len__(self) -> int:
        return self.count()

    def __contains__(self, name: str) -> bool:
        return self.exists(name)

    def __getitem__(self, name: str) -> DefinedName:
        if not self.exists(name):
            raise KeyError(name)
        return self.get(name)

    def __iter__(self) -> Iterator[DefinedName]:
        for dn in self._names:
            yield DefinedName(dn)

    def _resolve_sheet_id(self, sheet: Any) -> int:
        if isinstance(sheet, int):
            return sheet
        if hasattr(sheet, "index"):
            return int(sheet.index)
        if isinstance(sheet, str) and self._workbook is not None:
            names = list(self._workbook.sheetnames)
            try:
                return names.index(sheet)
            except ValueError as exc:
                raise KeyError(f"Sheet not found: {sheet!r}") from exc
        raise TypeError(
            "sheet must be a Worksheet, sheet title, or 0-based sheet index"
        )

    def __getattr__(self, name: str) -> Any:
        return getattr(self._names, name)
